skada/_subspace.py

```diff
@@ -366,3 +365,240 @@ def TransferComponentAnalysis(
+            (same as sample_domain).
+        weights : None
+            No weights are returned here.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain,
+            allow_multi_source=True,
+            allow_multi_target=True,
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        if np.array_equal(X_source, self.X_source_) and np.array_equal(
+            X_target, self.X_target_
+        ):
+            K = self._get_kernel_matrix(X_source, X_target)
+            X_ = K @ self.A_
+        else:
+            Ks = pairwise_kernels(X, self.X_source_, metric=self.kernel)
+            Kt = pairwise_kernels(X, self.X_target_, metric=self.kernel)
+            K = np.concatenate((Ks, Kt), axis=1)
+            X_ = K @ self.A_
+        return X_
+
+    def _get_mmd_matrix(self, ns, nt, sample_domain):
+        Mss = 1 / ns**2 * np.ones((ns, ns))
+        Mtt = 1 / nt**2 * np.ones((nt, nt))
+        Mst = -1 / (ns * nt) * np.ones((ns, nt))
+        M = np.block([[Mss, Mst], [Mst.T, Mtt]])
+        return M
+
+    def _get_kernel_matrix(self, X_source, X_target):
+        Kss = pairwise_kernels(X_source, metric=self.kernel)
+        Ktt = pairwise_kernels(X_target, metric=self.kernel)
+        Kst = pairwise_kernels(X_source, X_target, metric=self.kernel)
+        K = np.block([[Kss, Kst], [Kst.T, Ktt]])
+        return K
+
+    def l21_norm(self, A):
+        norm = 0
+        for i in range(A.shape[0]):
+            norm += np.linalg.norm(A[i])
+        return norm
+
+    def fit(self, X, y=None, sample_domain=None, **kwargs):
+        """Fit adaptation parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain,
+            allow_multi_source=True,
+            allow_multi_target=True,
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        if self.n_components is None:
+            n_components = min(min(X_source.shape), min(X_target.shape))
+        else:
+            n_components = self.n_components
+        self.random_state_ = check_random_state(self.random_state)
+
+        n = X.shape[0]
+        H = np.identity(n) - 1 / n * np.ones((n, n))
+        M = self._get_mmd_matrix(X_source.shape[0], X_target.shape[0], sample_domain)
+        self.X_source_ = X_source
+        self.X_target_ = X_target
+        K = self._get_kernel_matrix(X_source, X_target)
+        M /= np.linalg.norm(M, ord="fr" "o")
```

Comment:
is there a way for me to write "fro" without pre-commit raising an error? (because this is pretty ugly...)

Reply:
ok, I will do that